### DISCUSSION TITLE
덕스 패턴(Ducks pattern)

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#100] [프론트 기술 질문 공간]
덕스 패턴(Ducks pattern)이 무엇인가요?
덕스 패턴(Ducks pattern)에 대해 설명해주세요.

**Comment by WooGie911** [comment]
React에서 상태관리를 하기 위한 대표적인 라이브러리는 Redux입니다.
리덕스는 Store, Dispatch, Reducer, Action으로 되어 있어 하나의 상태를 저장하는 리덕스도 4개의 파일이 나오게되며 기존 리덕스에 하나의 액션을 추가하려면 3개의 파일을 수정해야 합니다. 
이 때문에 [Erik Rasmussen](https://github.com/erikras)님이 구조가 아닌, 기능을 기준으로 파일을 분리하는 것을 제안한 패턴입니다.

덕스 패턴은 여러 규칙을 가지는데 규칙은 아래와 같습니다.
- MUST export default a function called reducer()
-> 반드시 리듀서 함수를 default export해야 한다.

- MUST export its action creators as functions
-> 반드시 액션 생성 함수를 export해야 한다.

- MUST have action types in the form npm-module-or-app/reducer/ACTION_TYPE
-> 반드시 접두사를 붙인 형태로 액션 타입을 정의해야 한다. 

- MAY export its action types as UPPER_SNAKE_CASE, if an external reducer needs to listen for them, or if it is a published reusable library
-> (필수는 아닌데) 외부 리듀서가 모듈 내 액션 타입을 바라보고 있거나, 모듈이 재사용 가능한 라이브러리로 쓰이는 것이라면 액션 타입을 UPPER_SNAKE_CASE 형태로 이름 짓고 export 하면 된다.

**Comment by michoball** [comment]
앞서 ducks pattern의 조건에 대해 설명을 해주셔서 패턴이 생긴 이유와 예시를 첨부하도록 하겠습니다. 

통상적으로 redux를 이용해 특정 관심사에 대한 리덕스 기능을 만들때 use case에 따라 
`{액션타입, 액션, 리듀서}` 로 만들게 됩니다.  관심사별로 리덕스에 기능을 추가하게 되면 이 3가지를 지속적으로 추가해서 만들게 됩니다. 

예전에는 기능별로 폴더를 만들과 안에 세부적으로 3개의 파일을 만들어 관리를 했는데 그러다 기능의 변화가 생기면 거의 대부분의 경우 리듀서와 액션만이 서로 연관되어 수정이 필요했습니다. 

 서로 다른 파일을 오가며 관리하기에는 이 두개가 깊이 연관되어있어서 위 세가지 부분을 그냥 하나의 모듈로 묶어서 사용하는 것이 더 유리하다고 판단하였습니다. 

그래서 나온 것이 ducks pattern 입니다. 

앞서 설명하신 조건을 잘지키는 하나의 예시를 첨부하고 마치겠습니다. 

```javascript
// widgets.js

// Actions
const LOAD   = 'my-app/widgets/LOAD';
const CREATE = 'my-app/widgets/CREATE';
const UPDATE = 'my-app/widgets/UPDATE';
const REMOVE = 'my-app/widgets/REMOVE';

// Reducer
export default function reducer(state = {}, action = {}) {
  switch (action.type) {
    // do reducer stuff
    default: return state;
  }
}

// Action 생성자
export function loadWidgets() {
  return { type: LOAD };
}

export function createWidget(widget) {
  return { type: CREATE, widget };
}

export function updateWidget(widget) {
  return { type: UPDATE, widget };
}

export function removeWidget(widget) {
  return { type: REMOVE, widget };
}
```

자료위치 :  https://github.com/erikras/ducks-modular-redux

**Comment by javascriptbiggosoo** [comment]
덕스 패턴(Ducks pattern)은 리덕스(Redux) 모듈에서 액션 타입(Actions), 액션 생성 함수(Action Creators), 리듀서(Reducer)가 모두 들어있는 자바스크립트 파일을 의미합니다. 이렇게 하나의 파일에 작성하는 것을 덕스 패턴이라고 합니다. 리덕스 모듈에서 리듀서는 default export하고, 액션 생성 함수는 그냥 export합니다.
